In [2]:
import kagglehub

path = kagglehub.dataset_download(
    "irkaal/foodcom-recipes-and-reviews"
)

print("Path to dataset files:", path)

Path to dataset files: /Users/abbygoblick/.cache/kagglehub/datasets/irkaal/foodcom-recipes-and-reviews/versions/2


In [3]:
import os

os.listdir(path)

['reviews.csv', 'recipes.csv', 'recipes.parquet', 'reviews.parquet']

In [4]:
import pandas as pd
import os

recipes_path = os.path.join(path, "recipes.parquet")
reviews_path = os.path.join(path, "reviews.parquet")

recipes = pd.read_parquet(recipes_path)
reviews = pd.read_parquet(reviews_path)

In [5]:
print("Recipes:", recipes.shape)
print("Reviews:", reviews.shape)

Recipes: (522517, 28)
Reviews: (1401982, 8)


In [68]:
recipes.columns.tolist()

['RecipeId',
 'Name',
 'AuthorId',
 'AuthorName',
 'CookTime',
 'PrepTime',
 'TotalTime',
 'DatePublished',
 'Description',
 'Images',
 'RecipeCategory',
 'Keywords',
 'RecipeIngredientQuantities',
 'RecipeIngredientParts',
 'AggregatedRating',
 'ReviewCount',
 'Calories',
 'FatContent',
 'SaturatedFatContent',
 'CholesterolContent',
 'SodiumContent',
 'CarbohydrateContent',
 'FiberContent',
 'SugarContent',
 'ProteinContent',
 'RecipeServings',
 'RecipeYield',
 'RecipeInstructions']

In [7]:
reviews.columns.tolist()

['ReviewId',
 'RecipeId',
 'AuthorId',
 'AuthorName',
 'Rating',
 'Review',
 'DateSubmitted',
 'DateModified']

In [8]:
recipes.head()

,RecipeId,Name,AuthorId,AuthorName,CookTime,PrepTime,TotalTime,DatePublished,Description,Images,...,SaturatedFatContent,CholesterolContent,SodiumContent,CarbohydrateContent,FiberContent,SugarContent,ProteinContent,RecipeServings,RecipeYield,RecipeInstructions
0,38.0,Low-Fat Berry Blue Frozen Dessert,1533,Dancer,PT24H,PT45M,PT24H45M,1999-08-09 21:46:00+00:00,Make and share this Low-Fat Berry Blue Frozen ...,[https://img.sndimg.com/food/image/upload/w_55...,...,1.3,8.0,29.8,37.1,3.6,30.2,3.2,4.0,NaN,"[Toss 2 cups berries with sugar., Let stand fo..."
1,39.0,Biryani,1567,elly9812,PT25M,PT4H,PT4H25M,1999-08-29 13:12:00+00:00,Make and share this Biryani recipe from Food.com.,[https://img.sndimg.com/food/image/upload/w_55...,...,16.6,372.8,368.4,84.4,9.0,20.4,63.4,6.0,NaN,[Soak saffron in warm milk for 5 minutes and p...
2,40.0,Best Lemonade,1566,Stephen Little,PT5M,PT30M,PT35M,1999-09-05 19:52:00+00:00,This is from one of my first Good House Keepi...,[https://img.sndimg.com/food/image/upload/w_55...,...,0.0,0.0,1.8,81.5,0.4,77.2,0.3,4.0,NaN,"[Into a 1 quart Jar with tight fitting lid, pu..."
3,41.0,Carina's Tofu-Vegetable Kebabs,1586,Cyclopz,PT20M,PT24H,PT24H20M,1999-09-03 14:54:00+00:00,This dish is best prepared a day in advance to...,[https://img.sndimg.com/food/image/upload/w_55...,...,3.8,0.0,1558.6,64.2,17.3,32.1,29.3,2.0,4 kebabs,"[Drain the tofu, carefully squeezing out exces..."
4,42.0,Cabbage Soup,1538,Duckie067,PT30M,PT20M,PT50M,1999-09-19 06:19:00+00:00,Make and share this Cabbage Soup recipe from F...,[https://img.sndimg.com/food/image/upload/w_55...,...,0.1,0.0,959.3,25.1,4.8,17.7,4.3,4.0,NaN,"[Mix everything together and bring to a boil.,..."


In [9]:
reviews.head()

,ReviewId,RecipeId,AuthorId,AuthorName,Rating,Review,DateSubmitted,DateModified
0,2,992,2008,gayg msft,5,better than any you can get at a restaurant!,2000-01-25 21:44:00+00:00,2000-01-25 21:44:00+00:00
1,7,4384,1634,Bill Hilbrich,4,"I cut back on the mayo, and made up the differ...",2001-10-17 16:49:59+00:00,2001-10-17 16:49:59+00:00
2,9,4523,2046,Gay Gilmore ckpt,2,i think i did something wrong because i could ...,2000-02-25 09:00:00+00:00,2000-02-25 09:00:00+00:00
3,13,7435,1773,Malarkey Test,5,easily the best i have ever had. juicy flavor...,2000-03-13 21:15:00+00:00,2000-03-13 21:15:00+00:00
4,14,44,2085,Tony Small,5,An excellent dish.,2000-03-28 12:51:00+00:00,2000-03-28 12:51:00+00:00


In [10]:
print("Recipes:", recipes.shape)
print("Reviews:", reviews.shape)

Recipes: (522517, 28)
Reviews: (1401982, 8)


In [13]:
choc_chip_candidates = recipes[
    recipes["Name"].str.contains(
        "chocolate chip cookie",
        case=False,
        na=False
    )
].copy()

print("Candidates:", len(choc_chip_candidates))

Candidates: 1775


In [14]:
choc_chip_candidates[
    ["RecipeId", "Name", "RecipeCategory", "ReviewCount", "AggregatedRating"]
].head(50)

,RecipeId,Name,RecipeCategory,ReviewCount,AggregatedRating
168,218.0,Chocolate Chip Cookie Pizza,Dessert,NaN,NaN
1828,3307.0,Mrs. Fields Chocolate Chip Cookies,Drop Cookies,8.0,2.5
2109,3599.0,Best Basic Chocolate Chip Cookies,Drop Cookies,12.0,4.5
2415,3953.0,Oatmeal Chocolate Chip Cookies,Drop Cookies,19.0,5.0
2416,3954.0,Oatmeal Chocolate Chip Cookies (No Eggs),Drop Cookies,89.0,5.0
2487,4044.0,Ambrosia Oatmeal Chocolate Chip Cookies,Drop Cookies,7.0,3.5
2488,4045.0,Clementine's Oatmeal Chocolate Chip Cookies,Drop Cookies,8.0,4.5
2492,4050.0,Canadian Living's Best Chocolate Chip Cookies,Drop Cookies,85.0,5.0
2493,4051.0,Mike's Grand Marnier Chocolate Chip Cookies,Drop Cookies,2.0,4.0
2507,4065.0,Oatmeal Chocolate Chip Cookies II,Drop Cookies,68.0,5.0


In [15]:
choc_chip_candidates["RecipeCategory"].value_counts().head(30)

RecipeCategory
Drop Cookies              826
Dessert                   680
Chocolate Chip Cookies    144
Bar Cookie                 36
< 30 Mins                  13
Frozen Desserts            11
Cheesecake                  9
Pie                         8
Breakfast                   8
Candy                       6
Nuts                        4
< 60 Mins                   4
Quick Breads                3
Sweet                       3
Beverages                   2
Grains                      2
Ice Cream                   2
High Protein                2
Lunch/Snacks                1
Tarts                       1
Fruit                       1
Free Of...                  1
Peanut Butter               1
Kid Friendly                1
Canadian                    1
Toddler Friendly            1
Beans                       1
Shakes                      1
Breads                      1
< 15 Mins                   1
Name: count, dtype: int64

In [16]:
print("Total title candidates:", len(choc_chip_candidates))

Total title candidates: 1775


In [17]:
cookie_categories = [
    "Chocolate Chip Cookies",
    "Drop Cookies",
    "Dessert"
]

likely_cookies = choc_chip_candidates[
    choc_chip_candidates["RecipeCategory"].isin(cookie_categories)
].copy()

print("Likely cookie recipes:", len(likely_cookies))

Likely cookie recipes: 1650


In [18]:
high_confidence = choc_chip_candidates[
    choc_chip_candidates["RecipeCategory"] == "Chocolate Chip Cookies"
].copy()

drop_cookies = choc_chip_candidates[
    choc_chip_candidates["RecipeCategory"] == "Drop Cookies"
].copy()

other_candidates = choc_chip_candidates[
    ~choc_chip_candidates["RecipeCategory"].isin(
        ["Chocolate Chip Cookies", "Drop Cookies"]
    )
].copy()

print("Chocolate Chip Cookies category:", len(high_confidence))
print("Drop Cookies category:", len(drop_cookies))
print("Everything else:", len(other_candidates))
print("Total:", len(choc_chip_candidates))

Chocolate Chip Cookies category: 144
Drop Cookies category: 826
Everything else: 805
Total: 1775


In [19]:
choc_chip_candidates[
    [
        "RecipeId",
        "Name",
        "RecipeCategory",
        "ReviewCount",
        "AggregatedRating"
    ]
].sort_values(
    "ReviewCount",
    ascending=False
).head(50)

,RecipeId,Name,RecipeCategory,ReviewCount,AggregatedRating
17068,20408.0,Soft Chocolate Chip Cookies,Drop Cookies,509.0,5.0
75588,80470.0,Bakery Style Chewy Chocolate Chip Cookies,Drop Cookies,313.0,4.5
7004,10125.0,Soft Batch Chocolate Chip Cookies,Drop Cookies,243.0,5.0
95298,100727.0,"Thick, Soft, and Chewy Chocolate Chip Cookies",Drop Cookies,229.0,5.0
23372,26849.0,World's Best Chocolate Chip Cookies,Dessert,222.0,4.5
49241,53301.0,Chewy Double Chocolate Chip Cookies,Drop Cookies,221.0,5.0
47086,51104.0,Peanut Butter Chocolate Chip Cookies,Drop Cookies,186.0,5.0
34363,38041.0,My Favorite Chocolate Chip Cookies,Drop Cookies,186.0,5.0
95606,101041.0,Lisa's Swirled Chocolate Chip Cookies,Drop Cookies,183.0,5.0
111425,117317.0,Blue Ridge Mountains Chocolate Chip Cookies,Drop Cookies,168.0,5.0


In [20]:
exclude_terms = [
    "oatmeal",
    "peanut butter",
    "pumpkin",
    "banana",
    "bar",
    "cookie cake",
    "cookie pizza",
    "cookie dough",
    "raw",
    "gluten free",
    "gluten-free",
    "vegan",
    "eggless",
    "no eggs",
    "passover",
    "brownie",
    "white chocolate",
    "mint",
    "pudding",
    "cream cheese",
    "orange",
    "sour cream"
]

In [21]:
name_lower = choc_chip_candidates["Name"].str.lower()

exclude_mask = False

for term in exclude_terms:
    exclude_mask = exclude_mask | name_lower.str.contains(
        term,
        regex=False,
        na=False
    )

classic_candidates = choc_chip_candidates[
    ~exclude_mask
].copy()

print("Before filtering:", len(choc_chip_candidates))
print("After filtering:", len(classic_candidates))

Before filtering: 1775
After filtering: 1258


In [22]:
classic_candidates[
    [
        "RecipeId",
        "Name",
        "RecipeCategory",
        "ReviewCount",
        "AggregatedRating"
    ]
].sort_values(
    "ReviewCount",
    ascending=False
).head(50)

,RecipeId,Name,RecipeCategory,ReviewCount,AggregatedRating
17068,20408.0,Soft Chocolate Chip Cookies,Drop Cookies,509.0,5.0
75588,80470.0,Bakery Style Chewy Chocolate Chip Cookies,Drop Cookies,313.0,4.5
7004,10125.0,Soft Batch Chocolate Chip Cookies,Drop Cookies,243.0,5.0
95298,100727.0,"Thick, Soft, and Chewy Chocolate Chip Cookies",Drop Cookies,229.0,5.0
23372,26849.0,World's Best Chocolate Chip Cookies,Dessert,222.0,4.5
49241,53301.0,Chewy Double Chocolate Chip Cookies,Drop Cookies,221.0,5.0
34363,38041.0,My Favorite Chocolate Chip Cookies,Drop Cookies,186.0,5.0
95606,101041.0,Lisa's Swirled Chocolate Chip Cookies,Drop Cookies,183.0,5.0
111425,117317.0,Blue Ridge Mountains Chocolate Chip Cookies,Drop Cookies,168.0,5.0
10122,13307.0,Neiman-Marcus $250 Chocolate Chip Cookies Recipe,Dessert,156.0,5.0


In [24]:
classic_ids = classic_candidates["RecipeId"]

classic_reviews = reviews[
    reviews["RecipeId"].isin(classic_ids)
].copy()

print("Classic cookie recipes:", len(classic_candidates))
print("Associated reviews:", len(classic_reviews))

Classic cookie recipes: 1258
Associated reviews: 5651


In [25]:
review_counts = (
    classic_reviews
    .groupby("RecipeId")
    .size()
    .rename("actual_review_count")
)

classic_with_review_counts = classic_candidates.merge(
    review_counts,
    on="RecipeId",
    how="left"
)

classic_with_review_counts["actual_review_count"] = (
    classic_with_review_counts["actual_review_count"]
    .fillna(0)
    .astype(int)
)

classic_with_review_counts[
    ["Name", "ReviewCount", "actual_review_count"]
].sort_values(
    "actual_review_count",
    ascending=False
).head(20)

,Name,ReviewCount,actual_review_count
28,Soft Chocolate Chip Cookies,509.0,488
118,Bakery Style Chewy Chocolate Chip Cookies,313.0,294
8,Soft Batch Chocolate Chip Cookies,243.0,233
163,"Thick, Soft, and Chewy Chocolate Chip Cookies",229.0,221
39,World's Best Chocolate Chip Cookies,222.0,216
76,Chewy Double Chocolate Chip Cookies,221.0,214
53,My Favorite Chocolate Chip Cookies,186.0,182
165,Lisa's Swirled Chocolate Chip Cookies,183.0,171
201,Blue Ridge Mountains Chocolate Chip Cookies,168.0,162
12,Neiman-Marcus $250 Chocolate Chip Cookies Recipe,156.0,148


In [26]:
phenotype_terms = [
    "chewy",
    "chewier",
    "crispy",
    "crisp",
    "crunchy",
    "soft",
    "gooey",
    "cakey",
    "cake-like",
    "thick",
    "thin",
    "flat",
    "spread",
    "spreading",
    "dry",
    "moist",
    "dense"
]

for term in phenotype_terms:

    mask = (
        classic_reviews["Review"]
        .fillna("")
        .str.contains(
            term,
            case=False,
            regex=False
        )
    )

    n_reviews = mask.sum()

    n_recipes = (
        classic_reviews
        .loc[mask, "RecipeId"]
        .nunique()
    )

    print(
        f"{term:12} "
        f"reviews={n_reviews:5,} | "
        f"recipes={n_recipes:4,}"
    )

chewy        reviews=  869 | recipes= 220
chewier      reviews=   22 | recipes=  16
crispy       reviews=  240 | recipes= 116
crisp        reviews=  420 | recipes= 173
crunchy      reviews=  144 | recipes=  64
soft         reviews=  921 | recipes= 221
gooey        reviews=   79 | recipes=  41
cakey        reviews=   70 | recipes=  39
cake-like    reviews=   25 | recipes=  19
thick        reviews=  195 | recipes=  68
thin         reviews=1,084 | recipes= 286
flat         reviews=  402 | recipes= 140
spread       reviews=  224 | recipes= 106
spreading    reviews=   25 | recipes=  23
dry          reviews=  187 | recipes=  79
moist        reviews=  173 | recipes=  77
dense        reviews=   27 | recipes=  18


In [27]:
print("Recipes:", len(classic_candidates))
print("Reviews:", len(classic_reviews))
print(
    "Recipes with at least one review:",
    classic_reviews["RecipeId"].nunique()
)

Recipes: 1258
Reviews: 5651
Recipes with at least one review: 631


In [28]:
def sample_trait_reviews(trait, n=20):
    matches = classic_reviews[
        classic_reviews["Review"]
        .fillna("")
        .str.contains(
            trait,
            case=False,
            regex=False
        )
    ]

    print(f"Reviews containing '{trait}': {len(matches)}")

    return matches[
        ["RecipeId", "Rating", "Review"]
    ].sample(
        min(n, len(matches)),
        random_state=42
    )

In [29]:
sample_trait_reviews("chewy", 20)

Reviews containing 'chewy': 869


,RecipeId,Rating,Review
695333,80470,5,"These cookies are awesome! Very dense, slightl..."
134905,100727,5,Excellent! The very concise instructions were ...
882976,38041,5,These cookies are amazing! I am currently tryi...
135044,10289,4,These were very good--thick and crispy on the ...
1371249,48356,1,"Um they are very cake like, not soft and chewy..."
1244900,57336,5,I have tried MANY cookie recipes (including a ...
159859,76416,5,"Loved 'em, simple as that. Wonderfully chewy."
1299876,53301,5,WOW. These are delicious! They are exactly wha...
1033024,67068,5,Great recipe! I am weird about using less tha...
841577,10117,5,My daughter wanted to make chocolate chip cook...


In [30]:
pd.set_option("display.max_colwidth", None)

In [31]:
sample_trait_reviews("soft", 20)

Reviews containing 'soft': 921


,RecipeId,Rating,Review
567222,287336,5,"Delicious cookies! I have made these twice now. The first time I made them for my family and the second time I made them for work. Both my family and my co-workers enjoyed them. My cookies were soft, yet the rice krispies gave them a crunch that we all loved Thank you for posting this winning recipe! (Made for PAC - Spring 2008)"
654800,20408,5,"Oh boy these cookies are TERRIFIC!! I left out the almond extract and used butterscotch pudding. So good, soft and rich, they cooked up perfectly in the 12 minutes. I got 4 dozen good-sized cookies from my batch of dough. I will definitely use this recipe again - thanks for sharing it! Made for Fall PAC 2008"
893129,80470,3,These cookies were beautiful! BUT the flavor was not what I hoped. Too sweet for me. They were not soft and chewy but I did not cool them on the baking sheet so that might be the reason. I have no problem with crunchy cookies if the flavor is good. Like one reviewer said the flavor of the dough was not so great. I love cookie dough but this was not flavorful at all. I wish I had been advised that after baking they would not be golden brown because I probably over baked them looking for a golden color. If these had Toll House/Mrs. Field's flavor to go with the look they would have been a keeper. But unfortunately I can't say I will try again. 3 stars for the look and an ok flavor.
527566,250362,5,"I found my forever choc chip cookie recipe with this one. I am just not gifted in the baking department. I have tried many cookie recipes here and thankfully have always made out but not like I have with this one. I made subs too so having said that I thought for sure they wouldnt turn out. I used all organic ingredients; and used whole wheat flour instead of white. Used sugar in the raw (turbinado?) for the sugar. I left the cookies plain, then after they were spooned onto the cookie sheet i added carob chips in some, walnuts in others, organic dark chocolate chunks in others and some left plain. My 4yr old daughter, who helped to make them loved them too. Texture was great; chewey and soft and the flavor was right on as well; very very fresh tasting. Marney you were right; I wont be going back to another chocolate chip cookie again. Thank you so much for sharing this."
890319,20408,5,"I have made these several times and forgot to rate this recipe. It has always come out soft, chewy, and declicious no matter what kind of pudding I use. I have used chocolate fudge, regular chocolate, vanilla, and banana cream. Thanks Peach :)"
141941,57889,5,"My search has ended!! Thank you so much, luvfood. Chocolate chip is my husband's favorite. For the last 7 years I have tried to find a recipe for choc. chip cookies that stay soft, but don't spread out thin. This is it! These are just wonderful. Followed the recipe exactly. Mine took 11-12 minutes on cookie sheets."
842465,331287,5,"Wow! I made these cookies on vacation and my family devoured all 48 cookies in less than 24 hours! They raved that these were the best cookies ever! Very soft, moist, and full of chocolate chips. Wonderful recipe!"
1105875,20408,0,"I really liked these cookies. We do not eat a lot of cookies here so I made one batch, baked half of them then froze the other half in 'pucks' for cooking later. They stay soft and there is so many different things you can put in the cookies. Can't wait to try with different flavored pudding."
1244871,10125,5,"These are just perfect, perfectly soft that is. The pudding is the secret ingredient and somehow makes it so that the cookies stay soft until they are all gone. I keep pudding on hand just so my husband can make these cookies."
71664,67068,5,"Finally! A soft chocolate chip cookie that's really soft! This recipe truly lives up to it's name. It's Delicious and Soft! I did use margerine instead of butter, and chocolate chips instead of chunks, and made smaller cookies so ended up with 3 1/2 dozen. The only mistake I made was not doublin

In [32]:
sample_trait_reviews("crisp", 20)

Reviews containing 'crisp': 420


,RecipeId,Rating,Review
521622,22765,4,"My daughter and I left out the shortening and add in some rice crispies. Very good. They're still cooling in the kitchen, but we couldn't resist trying one."
1130783,218256,0,"These are the easiest, most delightful chocolate chip cookies I have ever made... In my opinion, the ideal cookie: crisp and golden brown on the outside, chewy and gooey on the inside. I have to say I am a tad confused about other reviews saying that they had so much trouble with these; mine came out to perfection. <br/><br/>I followed the recipe almost exactly, and used a stand mixer to combine all ingredients. I accidentally used 1/2 tsp. salt instead of 1/4, but they came out absolutely delicious! As recommended by another reviewer, I only used 1 1/4 cups chocolate chips, which is more than enough. I rolled the refrigerated cookie dough into 24 heaping tablespoon-sized balls and placed them on a dark, ungreased baking sheet, and actually baked for 15 minutes on the middle rack for a PERFECT cookie (found 8-10 minutes at an accurate 425? too short; the dough was still raw and pale). The key to getting them off the sheet onto a cooling rack is to have patience and let the cookies sit on the sheet for 2 minutes before attempting to transfer. Mine came right up, no problem. Delish! Definitely making them again..."
599876,30565,5,"These are so good! I love the combination of crispy and chewy and the recipe made a TON of cookies! I used only 1 cup of chocolate chips and it was plenty. I found that at 350, the cookies were browning a little too much on the bottoms, so I baked them at 335 degrees for about 10-11 minutes. So good! Thanks!"
1245400,20408,5,Almost perfect recipe. We cooked them on parchment paper on a cookie sheet. Which makes sure you get the good crisp edges. We also used a large scoop so we had the really good soft inside. I&#039;m going to have to play with the recipe to replace the pudding mix with cornstarch or cake flour because the flavor was just slightly off in my opinion.
1392714,100727,5,These are the BEST! I love that this recipe uses brown sugar instead of white. I also love the addition of crisco. It adds a little bit more crispness to the outside leaving the inside soft/chewy.
254914,53301,5,"These tasty little morcels should be given a 10+ star rating ! I haven't tasted anything so decadent and easy in a long time. As they were still warm from the oven each chocolate chip created an ooey goey chocolatey puddle with a slitely crispy exterior to each bite. You have to try these cookies. I had the neighborhood kids, My DH and DD ranting and raving over these ones. I am known as the cookie lady in my neighborhood and this one ranks on top. Be sure to have a cold glass of milk on hand... you'll need it. THANKS SO MUCH FOR AN AMAZING RECIPE !!!."
485279,39852,5,"These are good and chewy, not dry and crispy. They do make alot more than my blender could handle so towards the end of adding the six cups of flour I had to mix it by hand. Good cookies"
493014,80470,4,"I used this recipe when I forgot to take the butter out to soften for my usual recipe. We thought the flavor was good, but the cookies were more crisp than chewy. They disappeared quickly though! Thanks!"
127966,20408,5,A wonderfully soft and yet lightly crisp chocolate chip cookie. We all loved them. I omitted the almond extract only because they are going in a care package for a soldier in Iraq and I wasn't sure if he would like it. I think it sounds good though. Very pleased with how soft they are and am hoping they won't go *too* stale in the trip overseas. Great recipe!
250111,101041,5,Wonderful cookie recipe! I found that two packages (about 3 cups) of the swirled chips was perfect for us. The base had a wonderful flavor that stood up well to the sweetness of the chips. The edges were a little crispy and the middles a little chewy. Even DD's boyfriend proclaimed these REALLY good! A keeper for sure.


In [33]:
sample_trait_reviews("thick", 20)

Reviews containing 'thick': 195


,RecipeId,Rating,Review
1031310,358549,0,"Wow....hold on...I'm eating one of these as I type... Ok, phew!! Delicious!!! This was different from all of the other recipes that I have tried (even the recipes that have over 250 reviews at 4.5 starts) and this is by far the best. It does seem to need a touch more sugar, but overall...PERFECT. They turned out amazing! I followed the recipe without making any changes; however I did have to bake these for about 16 minutes. But that is totally fine with me... These were super easy to make since you don't have to wait for butter to come to room temperature! I typically don't like planning that far in advance. I'm an impulse baker and having to wait for butter to come to room temperature is a pain so another bonus by using this recipe! Try these! Next time I will add an additional 1/4 cup of sugar. These are going in my ""favorites"".<br/><br/>UPDATE: After these sat for a day... they were really hard and I could taste how prominent the salt was. Way too much salt! So... I'm now making these a 2nd time, but really tweaking this recipe. When made as-is, these cookies will get hard and they do not spread out when baking. For some I'm sure this is ok, but I like chewy/soft cookies...and trying to bake these thick cookies at 325 just doesn't cut it. You have to bake them longer to get them baked all the way through so in turn you are not getting soft cookies. Also, I don't really like baking cookies with shortening. So here's what I did this time... I like these much better!!! I used margarine instead of shortening... I used half the amount of salt and I added apple cider vinegar. For those of you who have never tried this - this will ensure your cookies are soft. This is a trick I learned from my mom who is an awesome baker."
270120,117317,5,Yummy! Words cannot express how excited I am about this recipe. The almond adds that little special something. The baking time was perfect. I added the extra flour to make my cookies puffy and thick. 10 stars all the way!!!
1141311,360266,5,"There is no other cookie recipe for me and the hub. After trying this Tyler Florence recipe, all other cookie recipes do not measure up. Thanks for posting DoubleA. I referenced this recipe often until I had it memorized from so many batches!<br/><br/>After making these cookies several times, these tweaks make the best cookies:<br/>1. Use Ghirardelli 60% bittersweet chocolate chips. The chocolate chips are shaped a little differently and melt a little differently so that you end up with ribbons of chocolate. You get a bit of chocolate in every bite. And the bittersweet chocolate keeps the cookie from being too sweet overall.<br/>2. Use unsalted butter, and, use just a pinch more salt than what is called for. I stick with kosher salt.<br/>3. After you mix your dough, use a 1/4 cup spring load ice cream scoop to pre-measure your dough into perfect little half-balls. Make the scoop absolutely level. Pop out into a large wax paper lined tupperware. Layering is okay. You end up with 16 cookie balls total, with about one spoonful of left over dough for the chef ;)<br/>4. Refrigerate your dough balls over night. After I started pre-measuring my cookies, I realized how awesome it was to have fresh cookies anytime I wanted them, rather than have half a batch of cookies go stale. And, I found that cookies from day before dough tasted better than same day cookies. I don't know why, they just do. The dough texture is totally different.<br/>5. Bake at 350, straight from fridge, for about 15 minutes. They should look just barely golden at the bottom and still a little doughy on top. They are a little thicker than a normal cookie and perfectly round. Let them sit on the cookie sheet for another 10-15 minutes. It will almost kill you, but I promise it is worth it. You end up with the perfect cookie - crisp on the bottom and practically still gooey in the center. They look, smell and taste perfect. So fabulous with milk."
839592,117

In [34]:
sample_trait_reviews("thin", 20)

Reviews containing 'thin': 1084


,RecipeId,Rating,Review
937506,20408,5,These were awesome cookies. I loved that they didn't flatten. I did use regular shortening and next time I think I will use the butter flavoring for extra flavor. Thanks for sharing a great recipe.
92285,10429,3,"These cookies are just okay. If I ever make them again I think I would add another egg. Also, 12 minutes in the oven would be far too long."
463899,100727,5,"This is the first time I made a soft chocolate chip cookie. Usually when I make them they are more crisp and crunchy. These were very good. Nice and soft and chewy. I like these much better than the thinner crispy chocolate chip cookies. I had 1/2 a cup of white chocolate chips that I wanted to use up, so I used the 1/2 cup of white chocolate chips and 1 cup of the semi-sweet chocolate chips. I also added a cup of chopped pecans. These cookies were excellent. I will be making my chocolate chip cookies this way from now on. Thanks for sharing such a wonderful recipe, Marg."
1327855,20408,5,"These are yummy! My whole family loved them and they were flying off the rack before they could cool. It made a lot, so they are getting extra treats in their lunch boxes. I think the vanilla pudding mix is a great addition in cookies."
338035,39641,5,"I cannot believe that this recipe has been on Zaar for 5 years and yet no one has ever reviewed it!! You all are missing out! These cookies are FANTASTIC! A 10-star recipe all the way. Absolutely everyone who has tried one has raved about them. I made them exactly as written, using walnuts and peanut butter chips. Even though I think some various combinations of nuts and chips might be fun, make sure you don't omit the butter brickle pieces; they really add a unique dimension to these cookies. I will definitely be making them again! I would suggest checking them at 9 minutes since in my oven they were a bit overdone at 10 minutes. Thank you so much for sharing this fantastic recipe! I'm only sorry I didn't find it sooner!"
128406,81016,5,"I've never made cookies where you mix everything all together in one bowl, so I did it my usual way.Using my Kitchen Aid I creamed the Crisco and sugar till fluffy,then added the eggs and vanilla, mixed that up and then added the flour and mixed up, then the choc. chips. I had miniature chips so only used 1 cup. I got 28 cookies using a 2"" cookie scoop. They did spread quite a bit but they turned out delicious.I baked for 14 minutes. Turned out nice and crisp, just the way we like them. Thanks for your mom's recipe."
624607,293380,4,These are wonderful cookies. I liked the texture of the cookie and they did keep soft for a day. I used semi-sweet chips just as that is what I had on hand but everything else remained the same. Made for PAC April 2008.
706954,332767,5,I really did think these were pretty great! I halved the recipe because I wanted to see how they'd come out and was not disappointed. I baked them 9 - 10 minutes so they came out soft in the middle and crispy on the sides. Also - I added 1 tsp of vanilla extract and 1 tsp of peppermint extract (instead of 2 tsp vanilla) so it was like biting into a mint-chocolate. Loved them and will be making these with several variations. Thanks for the recipe blucoat.
470628,258922,5,"Very tasty cookies! I used 2 cups milk chocolate chips and 2 cups white chocolate chips just for something different, but the cookies would have been deelish no matter what chips or nuts you added. A definate keeper! Thanks!"
836892,101041,4,"Perfect texture!!! I couldn't get to my usual chocolate chip cookie recipe since moving, so I looked for a highly rated one on here and found this one. The texture of these are exactly like m other ones, which I love, and while they taste really good, theres something about the flavour that isn't quite as good as my other recipe (thats the one star loss). \nI did make these with chunks of dark chocolate instead of chocolate chips though, very yummy!"


# Cookie Lab: Texture Phenotype Audit

## Goal

The objective of this analysis is to determine whether recipe reviews can be used to extract measurable cookie characteristics ("cookie phenotypes").

Instead of classifying cookies into one category, Cookie Lab treats cookies as a combination of physical traits:

- Softness
- Chewiness
- Crispness
- Thickness
- Spread / Thinness
- Moisture
- Cakiness
- Gooeyness

Reviews were analyzed to determine whether users naturally describe these characteristics and whether they provide enough signal for future NLP modeling.

---

# Dataset Overview

Source:
Food.com Recipes and Reviews Dataset

## Initial Dataset

Recipes:
- 522,517 recipes
- 28 attributes

Reviews:
- 1,401,982 reviews
- 8 attributes


## Chocolate Chip Cookie Subset

Recipes identified through:

- Recipe category
- Recipe title matching
- Chocolate chip cookie keywords

Final subset:

Recipes:
- 1,258

Reviews:
- 5,651

Recipes with at least one review:
- 631


---

# Chocolate Chip Cookie Identification

The dataset contained an existing category:


In [36]:
classic_candidates["RecipeCategory"].value_counts()

RecipeCategory
Drop Cookies              606
Dessert                   491
Chocolate Chip Cookies    108
Bar Cookie                  8
Frozen Desserts             8
< 30 Mins                   8
Pie                         7
Breakfast                   3
Quick Breads                2
Beverages                   2
Sweet                       2
< 60 Mins                   2
Cheesecake                  1
Tarts                       1
Kid Friendly                1
Grains                      1
Canadian                    1
Toddler Friendly            1
Ice Cream                   1
Nuts                        1
Shakes                      1
Candy                       1
< 15 Mins                   1
Name: count, dtype: int64

In [37]:
classic_candidates[
    classic_candidates["RecipeCategory"].isin([
        "Bar Cookie",
        "Frozen Desserts",
        "Pie",
        "Breakfast",
        "Quick Breads",
        "Beverages",
        "Cheesecake",
        "Tarts",
        "Ice Cream",
        "Shakes",
        "Candy"
    ])
][["Name", "RecipeCategory"]]

,Name,RecipeCategory
16742,Chocolate Chip Cookie Turtle,Bar Cookie
23254,Jumbo Chocolate Chip Cookie,Bar Cookie
42856,Chocolate Chip Cookie Gift Mix,Bar Cookie
69291,Fluffy Chocolate Chip Cookies,Quick Breads
92813,Chocolate Chip Cookie Milkshake,Frozen Desserts
93884,Chocolate Chip Cookie Pie,Pie
107421,Granola Chocolate Chip Cookies,Breakfast
121808,Chocolate Chip Cookie Pie,Pie
129535,Oats 'n Caramel Chocolate Chip Cookie Squares,Bar Cookie
143114,Chocolate Chip Cookie Martini,Beverages


In [38]:
classic_cookies_clean = classic_candidates[
    classic_candidates["RecipeCategory"].isin([
        "Drop Cookies",
        "Chocolate Chip Cookies",
        "Dessert"
    ])
]

print(classic_cookies_clean.shape)

(1205, 28)


In [39]:
classic_cookies_clean[
    ~classic_cookies_clean["Name"].str.contains(
        "cookie",
        case=False,
        na=False
    )
][["Name","RecipeCategory"]]

,Name,RecipeCategory


In [40]:
classic_cookies_clean["RecipeCategory"].value_counts()

RecipeCategory
Drop Cookies              606
Dessert                   491
Chocolate Chip Cookies    108
Name: count, dtype: int64

In [41]:
classic_cookies_clean[["Name", "RecipeCategory"]].sort_values("Name")

,Name,RecipeCategory
509605,#1 Chocolate Chip Cookies,Chocolate Chip Cookies
169322,$250 Chocolate Chip Cookies Recipe,Dessert
239908,$250 Chocolate Chip Cookies Recipe,Dessert
112890,$250.00 Chocolate Chip Cookies,Drop Cookies
495594,&quot;Indonesian&quot; Chocolate Chip Cookies,Chocolate Chip Cookies
...,...,...
159320,Zander's Chocolate Chip Cookies,Dessert
277079,Zippy Chocolate Chip Cookies,Drop Cookies
419446,Zoey's Healthier Fluffy Chocolate Chip Cookies,Chocolate Chip Cookies
243762,Zucchini Chocolate Chip Cookies,Drop Cookies


In [43]:
classic_cookies_clean[
    classic_cookies_clean["Name"].str.contains(
        "bar|cake|pie|brownie|cupcake|muffin|ice cream|sandwich|dough|mix|turtle|martini|drink",
        case=False,
        na=False
    )
][["Name","RecipeCategory"]]

,Name,RecipeCategory
40980,Chocolate Chip Cookie Mix,Drop Cookies
44551,Chunky Chocolate Chip Cookie Mix in a Jar,Dessert
44555,Chocolate Chip Cookie Mix in a Jar,Drop Cookies
67310,Chocolate Chip Cookie Mix in a Jar,Drop Cookies
93757,Decadent Chocolate Chip Cookies(Cake Mix),Drop Cookies
117770,Ultimate Chocolate Chip Cookies in a Jar Mix,Dessert
141061,Chocolate Chip Cookie Cupcakes,Dessert
160992,Kentucky Pie (Giant Chocolate Chip Cookie Pie With Nuts),Dessert
192581,"Cranberry, Bittersweet Chocolate Chip Cookies in a Jar Mix",Drop Cookies
192583,Chocolate Chocolate Chip Cookies in a Jar Mix,Drop Cookies


In [44]:
remove_words = [
    "mix",
    "jar",
    "cupcake",
    "cake",
    "cheesecake",
    "pie",
    "sandwich",
    "icebox",
    "poke"
]

pattern = "|".join(remove_words)

classic_cookies_final = classic_cookies_clean[
    ~classic_cookies_clean["Name"].str.contains(
        pattern,
        case=False,
        na=False
    )
]

print(classic_cookies_final.shape)

(1180, 28)


In [45]:
classic_cookies_final[["Name","RecipeCategory"]]

,Name,RecipeCategory
1828,Mrs. Fields Chocolate Chip Cookies,Drop Cookies
2109,Best Basic Chocolate Chip Cookies,Drop Cookies
2492,Canadian Living's Best Chocolate Chip Cookies,Drop Cookies
2493,Mike's Grand Marnier Chocolate Chip Cookies,Drop Cookies
2529,Chocolate Chip Cookies,Drop Cookies
...,...,...
521670,Scrumptious Chocolate Chip Cookies,Dessert
521671,Scrumptious Chocolate Chip Cookies,Dessert
521674,Lindsay's Chocolate Cafe Chocolate Chip Cookies,Dessert
522156,BEST EVER CHEWY CHOCOLATE CHIP COOKIES,Dessert


In [46]:
classic_cookies_final["Name"].duplicated().sum()

np.int64(330)

In [47]:
classic_cookies_final[
    classic_cookies_final["Name"].duplicated(keep=False)
][["Name","RecipeCategory"]].sort_values("Name")

,Name,RecipeCategory
169322,$250 Chocolate Chip Cookies Recipe,Dessert
239908,$250 Chocolate Chip Cookies Recipe,Dessert
494530,4 Ingredient Chocolate Chip Cookies,Chocolate Chip Cookies
512397,4 Ingredient Chocolate Chip Cookies,Dessert
29265,Absolutely the Best Chocolate Chip Cookies,Drop Cookies
...,...,...
243057,Yummy Chocolate Chip Cookies,Drop Cookies
336862,Yummy Chocolate Chip Cookies,Dessert
359318,Yummy Chocolate Chip Cookies,Drop Cookies
243762,Zucchini Chocolate Chip Cookies,Drop Cookies


In [48]:
duplicate_names = classic_cookies_final[
    classic_cookies_final["Name"].duplicated(keep=False)
]["Name"]

classic_cookies_final[
    classic_cookies_final["Name"].isin(duplicate_names)
][["RecipeId", "Name", "AuthorName"]].sort_values("Name")

,RecipeId,Name,AuthorName
169322,177187.0,$250 Chocolate Chip Cookies Recipe,SaraC
239908,249746.0,$250 Chocolate Chip Cookies Recipe,Rissy
494530,512710.0,4 Ingredient Chocolate Chip Cookies,Bakers 5th Rush
512397,530937.0,4 Ingredient Chocolate Chip Cookies,Neriah M.
29265,32852.0,Absolutely the Best Chocolate Chip Cookies,Carrie Ann
...,...,...,...
243057,252982.0,Yummy Chocolate Chip Cookies,lrl418
336862,349520.0,Yummy Chocolate Chip Cookies,See-Jay
359318,372470.0,Yummy Chocolate Chip Cookies,Meghan lt3
243762,253705.0,Zucchini Chocolate Chip Cookies,mommaj


### Checked duplicates, inspected ID to make sure they were not the same recipes


In [49]:
classic_cookies_final[
    classic_cookies_final["Name"] == "Yummy Chocolate Chip Cookies"
][[
    "RecipeId",
    "Name",
    "RecipeIngredientParts",
    "RecipeInstructions"
]]

,RecipeId,Name,RecipeIngredientParts,RecipeInstructions
243057,252982.0,Yummy Chocolate Chip Cookies,"[all-purpose flour, baking soda, salt, butter, granulated sugar, brown sugar, vanilla extract, eggs]","[Preheat oven to 375., Beat together butter and sugar until creamy., Add Vanilla., Beat in Eggs one at a time., In a seperate bowl, combine flour, baking soda and salt in small bowl., Slowly add the flour mixture into the wet mixture., Stir in choc.chips. (Add nuts at this point too if you'd like)., Roll dough into little balls, about 1 tablespoon each, and place onto ungreased baking sheets., Bake for 11 minutes or until golden brown., Cool on baking sheets for 2 minutes; remove to wire racks to cool completely., ENJOY!]"
336862,349520.0,Yummy Chocolate Chip Cookies,"[vegetable shortening, white sugar, white sugar, brown sugar, brown sugar, vanilla, water, egg, flour, salt, baking soda]","[Preheat oven to 375 degrees., Beat shortening in mixer until fluffy., Add both sugars and beat for several minutes; until light and fluffy., Add vanilla, water, and egg. Beat until incorporated., In a seperate bowl, mix together the flour, salt, and baking soda., SLOWLY add the dry ingredients to the wet mixture and stir on low for several seconds., Add the chocolate chips., Using a scooper, place each ball of dough onto a cookiesheet., Bake for 8-10 minutes; or until you stick a toothpick in a cookie and it comes out clean.]"
359318,372470.0,Yummy Chocolate Chip Cookies,"[butter, brown sugar, white sugar, vanilla, egg, flour, baking soda, salt, cornstarch]","[Preheat oven to 325 degrees F., Cream together butter, sugar and vanilla. Add egg., In a seperate bowl, sift together flour, soda, salt and starch., Mix into creamed mixture., Add chocolate chips., Drop by tablespoons on a wax or parchment paper lined cookie sheet., Bake for about 12 minutes, or until golden brown., Let cool on the cookie sheet for about a minute. Then transfer to cookie sheet for further cooling.]"


In [53]:
%whos

Variable                     Type         Data/Info
---------------------------------------------------
choc_chip_candidates         DataFrame    Shape: (1775, 28)
classic_candidates           DataFrame    Shape: (1258, 28)
classic_cookie_reviews       DataFrame    Shape: (5515, 35)
classic_cookies_clean        DataFrame    Shape: (1205, 28)
classic_cookies_final        DataFrame    Shape: (1180, 28)
classic_ids                  Series       Shape: (1258,)
classic_reviews              DataFrame    Shape: (5651, 8)
classic_with_review_counts   DataFrame    Shape: (1258, 29)
cookie_categories            list         n=3
drop_cookies                 DataFrame    Shape: (826, 28)
duplicate_names              Series       Shape: (421,)
exclude_mask                 Series       Shape: (1775,)
exclude_terms                list         n=22
high_confidence              DataFrame    Shape: (144, 28)
kagglehub                    module       <module 'kagglehub' from <...>s/kagglehub/__init__.py'

In [51]:
classic_cookie_reviews.head()

,ReviewId,RecipeId,AuthorId_x,AuthorName_x,Rating,Review,DateSubmitted,DateModified,Name,AuthorId_y,...,SaturatedFatContent,CholesterolContent,SodiumContent,CarbohydrateContent,FiberContent,SugarContent,ProteinContent,RecipeServings,RecipeYield,RecipeInstructions
0,19,13307,2046,Gay Gilmore ckpt,5,"chewy goodness, not crispy at all. i even threw in craisins and left the oatmeal whole, and they were great.",2000-05-21 16:59:00+00:00,2000-05-21 16:59:00+00:00,Neiman-Marcus $250 Chocolate Chip Cookies Recipe,20914,...,4.5,18.5,127.6,22.8,1.5,13.9,2.5,NaN,96 cookies,"[Measure oatmeal and blend in a blender to a fine powder., Cream the butter and both sugars., Add eggs and vanilla., Mix together with flour, oatmeal, salt, baking powder and soda., Add chocolate chips, grated Hershey Bar and nuts., Roll into 1 inch balls and place 2-inches apart on a cookie sheet., Bake for 10 minutes at 375° or until golden.]"
1,253,13307,4200,LC Liu,5,"I received this ""hoax"" email forward about 4 years ago, but never attempted the recipe. A friend of the family made it for us just a few weeks ago and it tastes incredible! I can't believe I held on to this recipe for so long!!",2001-01-02 16:32:03+00:00,2001-01-02 16:32:03+00:00,Neiman-Marcus $250 Chocolate Chip Cookies Recipe,20914,...,4.5,18.5,127.6,22.8,1.5,13.9,2.5,NaN,96 cookies,"[Measure oatmeal and blend in a blender to a fine powder., Cream the butter and both sugars., Add eggs and vanilla., Mix together with flour, oatmeal, salt, baking powder and soda., Add chocolate chips, grated Hershey Bar and nuts., Roll into 1 inch balls and place 2-inches apart on a cookie sheet., Bake for 10 minutes at 375° or until golden.]"
2,585,4051,7209,Jazzy,4,"This makes a lovely chocolate chip cookie, but I couldn't taste the Grand Marnier. Perhaps adding a bit of orange zest would enhance the orange flavor.",2001-03-29 17:11:32+00:00,2001-03-29 17:11:32+00:00,Mike's Grand Marnier Chocolate Chip Cookies,1556,...,2.8,10.3,100.6,20.1,1.0,12.4,1.8,NaN,36 cookies,"[Cream margarine, sugars, vanilla and Grand Marnier in a large mixing bowl., Add eggs, beating well after each addition., Add baking soda and flour and stir until blended., Stir in oats, then chocolate chips., Drop by the tablespoonfuls 5 cm apart onto ungreased cookie sheets., Bake at 300°F for 5-10 minutes or until done. Allow to cool somewhat before removing from cookie sheet.]"
3,586,13307,7230,Sadie Schilk,5,"This recipe is awesome! I tried it out of curiosity, brought the cookies to work, and everyone loved them. They fought over who was going to eat the last one. It makes about a million cookies, so unless you are sharing it with a large number of people, cut the recipe in half, or you will have more cookies than you know what to do with!",2001-03-29 17:11:32+00:00,2001-03-29 17:11:32+00:00,Neiman-Marcus $250 Chocolate Chip Cookies Recipe,20914,...,4.5,18.5,127.6,22.8,1.5,13.9,2.5,NaN,96 cookies,"[Measure oatmeal and blend in a blender to a fine powder., Cream the butter and both sugars., Add eggs and vanilla., Mix together with flour, oatmeal, salt, baking powder and soda., Add chocolate chips, grated Hershey Bar and nuts., Roll into 1 inch balls and place 2-inches apart on a cookie sheet., Bake for 10 minutes at 375° or until golden.]"
4,630,3307,7673,Rachel Merrell,1,"dough is too thin, the cookies melted together and they came out really thin. Flavor was ok.",2001-04-04 16:51:30+00:00,2001-04-04 16:51:30+00:00,Mrs. Fields Chocolate Chip Cookies,1763,...,4.3,21.7,66.3,14.9,0.6,11.9,1.0,NaN,42 cookies,"[Heat oven to 300°F., Combine flour, baking soda and salt in medium bowl using whisk; set aside., Mix sugars in bowl with an electric mixer on medium speed., Add butter; beat until grainy paste forms., Beat in eggs and vanilla on medium speed until light and fluffy., Add flour mixture and chocolate chips., Beat on low speed just until mixed., Do not over mix., Drop dough by rounded tablespoons onto ungreased baking sheets, 2-inches apart. Bake

In [54]:
classic_cookie_reviews.head()

,ReviewId,RecipeId,AuthorId_x,AuthorName_x,Rating,Review,DateSubmitted,DateModified,Name,AuthorId_y,...,SaturatedFatContent,CholesterolContent,SodiumContent,CarbohydrateContent,FiberContent,SugarContent,ProteinContent,RecipeServings,RecipeYield,RecipeInstructions
0,19,13307,2046,Gay Gilmore ckpt,5,"chewy goodness, not crispy at all. i even threw in craisins and left the oatmeal whole, and they were great.",2000-05-21 16:59:00+00:00,2000-05-21 16:59:00+00:00,Neiman-Marcus $250 Chocolate Chip Cookies Recipe,20914,...,4.5,18.5,127.6,22.8,1.5,13.9,2.5,NaN,96 cookies,"[Measure oatmeal and blend in a blender to a fine powder., Cream the butter and both sugars., Add eggs and vanilla., Mix together with flour, oatmeal, salt, baking powder and soda., Add chocolate chips, grated Hershey Bar and nuts., Roll into 1 inch balls and place 2-inches apart on a cookie sheet., Bake for 10 minutes at 375° or until golden.]"
1,253,13307,4200,LC Liu,5,"I received this ""hoax"" email forward about 4 years ago, but never attempted the recipe. A friend of the family made it for us just a few weeks ago and it tastes incredible! I can't believe I held on to this recipe for so long!!",2001-01-02 16:32:03+00:00,2001-01-02 16:32:03+00:00,Neiman-Marcus $250 Chocolate Chip Cookies Recipe,20914,...,4.5,18.5,127.6,22.8,1.5,13.9,2.5,NaN,96 cookies,"[Measure oatmeal and blend in a blender to a fine powder., Cream the butter and both sugars., Add eggs and vanilla., Mix together with flour, oatmeal, salt, baking powder and soda., Add chocolate chips, grated Hershey Bar and nuts., Roll into 1 inch balls and place 2-inches apart on a cookie sheet., Bake for 10 minutes at 375° or until golden.]"
2,585,4051,7209,Jazzy,4,"This makes a lovely chocolate chip cookie, but I couldn't taste the Grand Marnier. Perhaps adding a bit of orange zest would enhance the orange flavor.",2001-03-29 17:11:32+00:00,2001-03-29 17:11:32+00:00,Mike's Grand Marnier Chocolate Chip Cookies,1556,...,2.8,10.3,100.6,20.1,1.0,12.4,1.8,NaN,36 cookies,"[Cream margarine, sugars, vanilla and Grand Marnier in a large mixing bowl., Add eggs, beating well after each addition., Add baking soda and flour and stir until blended., Stir in oats, then chocolate chips., Drop by the tablespoonfuls 5 cm apart onto ungreased cookie sheets., Bake at 300°F for 5-10 minutes or until done. Allow to cool somewhat before removing from cookie sheet.]"
3,586,13307,7230,Sadie Schilk,5,"This recipe is awesome! I tried it out of curiosity, brought the cookies to work, and everyone loved them. They fought over who was going to eat the last one. It makes about a million cookies, so unless you are sharing it with a large number of people, cut the recipe in half, or you will have more cookies than you know what to do with!",2001-03-29 17:11:32+00:00,2001-03-29 17:11:32+00:00,Neiman-Marcus $250 Chocolate Chip Cookies Recipe,20914,...,4.5,18.5,127.6,22.8,1.5,13.9,2.5,NaN,96 cookies,"[Measure oatmeal and blend in a blender to a fine powder., Cream the butter and both sugars., Add eggs and vanilla., Mix together with flour, oatmeal, salt, baking powder and soda., Add chocolate chips, grated Hershey Bar and nuts., Roll into 1 inch balls and place 2-inches apart on a cookie sheet., Bake for 10 minutes at 375° or until golden.]"
4,630,3307,7673,Rachel Merrell,1,"dough is too thin, the cookies melted together and they came out really thin. Flavor was ok.",2001-04-04 16:51:30+00:00,2001-04-04 16:51:30+00:00,Mrs. Fields Chocolate Chip Cookies,1763,...,4.3,21.7,66.3,14.9,0.6,11.9,1.0,NaN,42 cookies,"[Heat oven to 300°F., Combine flour, baking soda and salt in medium bowl using whisk; set aside., Mix sugars in bowl with an electric mixer on medium speed., Add butter; beat until grainy paste forms., Beat in eggs and vanilla on medium speed until light and fluffy., Add flour mixture and chocolate chips., Beat on low speed just until mixed., Do not over mix., Drop dough by rounded tablespoons onto ungreased baking sheets, 2-inches apart. Bake

In [56]:
classic_cookie_reviews["RecipeIngredientParts"] = (
    classic_cookie_reviews["RecipeIngredientParts"]
    .astype(str)
)

classic_cookie_reviews["RecipeInstructions"] = (
    classic_cookie_reviews["RecipeInstructions"]
    .astype(str)
)

In [57]:
classic_cookie_recipe_level = (
    classic_cookie_reviews
    .groupby(
        [
            "RecipeId",
            "Name",
            "RecipeIngredientParts",
            "RecipeInstructions"
        ]
    )
    .agg(
        num_reviews=("Review", "count"),
        avg_rating=("Rating", "mean"),
        all_reviews=("Review", " ".join)
    )
    .reset_index()
)

In [58]:
classic_cookie_recipe_level.shape

(594, 7)

In [59]:
classic_cookie_recipe_level.head()

RecipeId                                           Name  \
0      3307             Mrs. Fields Chocolate Chip Cookies   
1      3599              Best Basic Chocolate Chip Cookies   
2      4050  Canadian Living's Best Chocolate Chip Cookies   
3      4051    Mike's Grand Marnier Chocolate Chip Cookies   
4      4092                         Chocolate Chip Cookies   

                                                                                                                                         RecipeIngredientParts  \
0                                                    ['all-purpose flour' 'baking soda' 'salt' 'dark brown sugar' 'sugar'\n 'butter' 'eggs' 'vanilla extract']   
1                      ['shortening' 'butter' 'brown sugar' 'granulated sugar' 'baking soda'\n 'salt' 'eggs' 'vanilla' 'all-purpose flour' 'walnuts' 'pecans']   
2  ['butter' 'shortening' 'granulated sugar' 'brown sugar' 'eggs' 'vanilla'\n 'all-purpose flour' 'baking soda' 'salt' 'chocolate chips' 'walnuts'\n 'pecans']   
3                             ['margarine' 'brown sugar' 'granulated sugar' 'vanilla' 'Grand Marnier'\n 'eggs' 'all-purpose flour' 'baking soda' 'quick oats']   
4                               ['butter' 'caster sugar' 'brown sugar' 'egg' 'vanilla extract'\n 'plain flour' 'salt' 'bicarbonate of soda' 'chocolate chips']   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                RecipeInstructions  \
0                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          

In [60]:
texture_words = [
    "chewy",
    "chewier",
    "crispy",
    "crisp",
    "crunchy",
    "soft",
    "gooey",
    "cakey",
    "cake-like",
    "thick",
    "thin",
    "flat",
    "spread",
    "spreading",
    "dry",
    "moist",
    "dense"
]

In [61]:
for word in texture_words:
    review_mentions = (
        classic_cookie_recipe_level["all_reviews"]
        .str.lower()
        .str.count(word)
        .sum()
    )
    
    recipe_mentions = (
        classic_cookie_recipe_level["all_reviews"]
        .str.lower()
        .str.contains(word)
        .sum()
    )
    
    print(f"{word:10} reviews={review_mentions:5.0f} | recipes={recipe_mentions}")

chewy      reviews=  918 | recipes=215
chewier    reviews=   23 | recipes=16
crispy     reviews=  248 | recipes=113
crisp      reviews=  454 | recipes=169
crunchy    reviews=  153 | recipes=64
soft       reviews= 1033 | recipes=218
gooey      reviews=   79 | recipes=39
cakey      reviews=   73 | recipes=38
cake-like  reviews=   26 | recipes=19
thick      reviews=  208 | recipes=66
thin       reviews= 1301 | recipes=270
flat       reviews=  440 | recipes=138
spread     reviews=  237 | recipes=101
spreading  reviews=   22 | recipes=20
dry        reviews=  201 | recipes=75
moist      reviews=  175 | recipes=74
dense      reviews=   28 | recipes=18


In [62]:
for word in texture_words:
    column_name = word.replace("-", "_")
    
    classic_cookie_recipe_level[column_name] = (
        classic_cookie_recipe_level["all_reviews"]
        .str.lower()
        .str.contains(word)
        .astype(int)
    )

In [63]:
classic_cookie_recipe_level[
    [
        "Name",
        "soft",
        "chewy",
        "thin",
        "crispy",
        "thick"
    ]
].head(20)

,Name,soft,chewy,thin,crispy,thick
0,Mrs. Fields Chocolate Chip Cookies,0,0,1,0,0
1,Best Basic Chocolate Chip Cookies,1,0,1,0,0
2,Canadian Living's Best Chocolate Chip Cookies,1,1,1,1,1
3,Mike's Grand Marnier Chocolate Chip Cookies,1,0,0,0,0
4,Chocolate Chip Cookies,1,0,1,0,0
5,Decadent Chocolate Chip Cookies,1,1,1,1,1
6,Crispy Chewy Chocolate Chip Cookies,1,1,1,1,0
7,Best Ever Chocolate Chip Cookies,1,1,1,1,0
8,Soft Batch Chocolate Chip Cookies,1,1,1,1,1
9,The Ultimate Chocolate Chip Cookies!,0,1,0,1,1


In [66]:
classic_cookie_recipe_level.to_csv(
    "classic_cookie_texture_dataset.csv",
    index=False
)

In [67]:
print(classic_cookie_recipe_level)


     RecipeId                                           Name  \
0        3307             Mrs. Fields Chocolate Chip Cookies   
1        3599              Best Basic Chocolate Chip Cookies   
2        4050  Canadian Living's Best Chocolate Chip Cookies   
3        4051    Mike's Grand Marnier Chocolate Chip Cookies   
4        4092                         Chocolate Chip Cookies   
..        ...                                            ...   
589    522478         BEST-EVER Chewy Chocolate Chip Cookies   
590    522705                         Chocolate Chip Cookies   
591    530178            Paleo Almond Chocolate Chip Cookies   
592    530828                         Chocolate Chip Cookies   
593    533923          Ultimate Chewy Chocolate Chip Cookies   

                                                                                                                                                    RecipeIngredientParts  \
0                                                         

# Classic Chocolate Chip Cookie Dataset Cleaning

## Objective

The goal of this project is to create a machine learning dataset that can predict chocolate chip cookie texture (such as soft, chewy, thin, or crispy) based on recipe characteristics including ingredients and instructions.

To create this dataset, the original recipe database was filtered to isolate classic chocolate chip cookie recipes and then combined with user reviews to generate texture labels.

---

# 1. Initial Dataset

The original dataset contained:

- **522,517 recipes**
- **1,401,982 reviews**

Each recipe contained information including:

- Recipe ID
- Recipe name
- Recipe category
- Ingredients
- Instructions
- Author information
- Recipe metadata

---

# 2. Identify Chocolate Chip Cookie Candidates

Recipes were first filtered based on recipe names containing chocolate chip cookie-related terms.

This created:

`choc_chip_candidates`

**1,775 recipes**

This dataset contained both true chocolate chip cookie recipes and unrelated recipes that included the phrase "chocolate chip cookie."

Examples of unrelated recipes included:

- Chocolate Chip Cookie Pie
- Chocolate Chip Cookie Ice Cream Cake
- Chocolate Chip Cookie Cheesecake
- Chocolate Chip Cookie Milkshake

---

# 3. Remove Non-Cookie Recipes

The candidate dataset was cleaned to remove recipes that were not traditional chocolate chip cookies.

Recipes were removed based on:

- Recipe categories
- Recipe names
- Keywords indicating non-cookie desserts

Examples of removed categories:

- Bar Cookie
- Pie
- Frozen Desserts
- Cheesecake
- Ice Cream
- Beverages
- Cakes
- Sandwiches
- Cookie Mixes

After removing unrelated recipes:

`classic_candidates`

**1,258 recipes**

---

# 4. Refine Classic Chocolate Chip Cookie Recipes

Additional cleaning was performed to create a more representative classic chocolate chip cookie dataset.

This produced:

`classic_cookies_clean`

**1,205 recipes**

Duplicate recipe names were then investigated.

A total of:

**421 duplicate recipe names**

were identified.

Duplicates were reviewed because identical recipe names could represent different recipes from different authors.

After duplicate analysis, the final cleaned recipe dataset was:

`classic_cookies_final`

**1,180 recipes**

This dataset contains:

- Recipe ID
- Recipe name
- Recipe category
- Ingredients
- Instructions
- Author information
- Recipe metadata

---

# 5. Merge Recipes With Reviews

The cleaned recipe dataset was merged with the review dataset using:

`RecipeId`

An inner join was used so that only recipes with available reviews were retained.

The resulting dataset:

`classic_cookie_reviews`

contained:

**5,515 review rows**

Each row represented:

- One review
- The associated cookie recipe
- Recipe information

---

# 6. Convert Review-Level Data Into Recipe-Level Data

Because each recipe could have many reviews, the review-level dataset was converted into a recipe-level dataset.

Reviews were grouped by recipe and combined.

For each recipe:

- All reviews were combined into one text field
- The number of reviews was calculated
- The average rating was calculated

The final recipe-level dataset:

`classic_cookie_recipe_level`

contained:

**594 recipes**

Each row represents one cookie recipe.

Schema:

| Column | Description |
|---|---|
| RecipeId | Unique recipe identifier |
| Name | Recipe name |
| RecipeIngredientParts | List of recipe ingredients |
| RecipeInstructions | Baking instructions |
| num_reviews | Number of reviews associated with the recipe |
| avg_rating | Average review rating |
| all_reviews | All reviews combined into one text field |

---

# 7. Final Dataset Purpose

The final dataset will be used to predict cookie texture.

The recipe information will serve as model inputs:

- Ingredients
- Instructions
- Recipe characteristics

The combined reviews will be used to generate texture labels.

Example:

Reviews:

> "These cookies are soft, chewy, and stay thick."

Converted into labels:

| Texture | Label |
|---|---|
| Soft | 1 |
| Chewy | 1 |
| Thick | 1 |
| Crispy | 0 |

These labels will become the prediction targets for machine learning models.

---

# Dataset Pipeline Summary

```
Raw Recipe Dataset
        |
        ↓
Chocolate Chip Cookie Candidates
(1,775 recipes)
        |
        ↓
Remove Non-Cookie Recipes
        |
        ↓
Classic Cookie Recipes
(1,258 recipes)
        |
        ↓
Clean + Remove Duplicates
        |
        ↓
classic_cookies_final
(1,180 recipes)
        |
        ↓
Merge With Reviews
        |
        ↓
classic_cookie_reviews
(5,515 reviews)
        |
        ↓
Aggregate Reviews by Recipe
        |
        ↓
classic_cookie_recipe_level
(594 recipes)
```

---

# Completed Steps

✅ Loaded recipe and review datasets  
✅ Identified chocolate chip cookie recipes  
✅ Removed unrelated desserts  
✅ Cleaned classic cookie recipes  
✅ Investigated duplicate recipes  
✅ Created final recipe dataset  
✅ Merged recipes with reviews  
✅ Aggregated reviews into recipe-level data  
✅ Prepared dataset for texture label generation and machine learning  

---

# Next Steps

The next notebook will focus on:

1. Exploratory texture analysis
2. Extracting texture labels from reviews:
   - Soft
   - Chewy
   - Thin
   - Crispy
   - Thick
   - Gooey
   - Cakey

3. Feature engineering:
   - Ingredient analysis
   - Recipe instruction analysis

4. Machine learning models to predict cookie texture